lib/src/data_stream/stream_writer.dart

```diff
@@ -0,0 +1,108 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:typed_data';
+
+import 'package:fixnum/fixnum.dart';
+
+import 'package:livekit_client/src/core/engine.dart';
+import 'package:livekit_client/src/types/other.dart';
+import '../proto/livekit_models.pb.dart' as lk_models;
+import '../types/data_stream.dart';
+
+class BaseStreamWriter<T, InfoType extends BaseStreamInfo> {
+  final StreamWriter<T> writableStream;
+  Function()? onClose;
+
+  final InfoType info;
+
+  BaseStreamWriter(
+      {required this.writableStream, required this.info, this.onClose});
+
+  Future<void> write(T chunk) async {
+    return writableStream.write(chunk);
+  }
+
+  Future<void> close() async {
+    await writableStream.close();
+    this.onClose?.call();
+  }
+}
+
+class TextStreamWriter extends BaseStreamWriter<String, TextStreamInfo> {
+  TextStreamWriter(
+      {required super.writableStream,
+      required super.info,
+      required super.onClose});
+}
+
+class BinaryStreamWriter extends BaseStreamWriter<Uint8List, ByteStreamInfo> {
```

Comment:
rename to `ByteStreamWriter` (I had this wrong in the JS implementation, sorry!)